### PR TITLE
RESTWS-833: Add REST resource for Implementation ID management

### PR DESCRIPTION
**Pull request by BartlomiejRasztabiga**
## Description of what I changed
Added ImplementationIdController2_0 for Implementation ID management.

## Issue I worked on
see https://issues.openmrs.org/browse/RESTWS-833
PR with documentation: https://github.com/openmrs/openmrs-contrib-rest-api-docs/pull/163

## Checklist: I completed these to help reviewers :)
- [X] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [X] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [X] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [X] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [X] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [X] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/ImplementationIdController2_0.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+	public Object getCurrentConfiguration() {
+		AdministrationService administrationService = Context.getAdministrationService();
+		ImplementationId implementationId = administrationService.getImplementationId();
+
+		return ConversionUtil.convertToRepresentation(implementationId, Representation.FULL);
+	}
+
+	@RequestMapping(method = RequestMethod.POST)
+	@ResponseStatus(HttpStatus.OK)
+	public void updateCurrentConfiguration(@RequestBody ImplementationId implementationId) {
```
Comment (HerbertYiga) [inline omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/ImplementationIdController2_0.java:47]: Thanks @BartlomiejRasztabiga  looking good to me cc @ibacher  @dkayiwa

File: omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/ImplementationIdController2_0Test.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+		assertEquals("name", PropertyUtils.getProperty(result, "name"));
+		assertEquals("test", PropertyUtils.getProperty(result, "implementationId"));
+		assertEquals("description", PropertyUtils.getProperty(result, "description"));
+		assertEquals("test", PropertyUtils.getProperty(result, "passphrase"));
+	}
+
+	@Test
+	public void shouldUpdateCurrentConfiguration() throws Exception {
+		// make POST call
+		String json = "{\"name\": \"test\",\"description\": \"test\",\"implementationId\": \"test\",\"passphrase\": \"test\"}";
```
Comment (dkayiwa) [inline omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/ImplementationIdController2_0Test.java:66]: Would be better to use different values for each field.

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/ImplementationIdController2_0.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+		if (exceptions.hasErrors()) {
+			throw new IllegalRequestException(exceptions);
+		}
+
+		try {
+			administrationService.setImplementationId(implementationId);
+		}
+		catch (APIException e) {
+			throw new IllegalRequestException(e);
```
Comment (dkayiwa) [inline omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/ImplementationIdController2_0.java:61]: How do you tell that this is an illegal request?

File: omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/ImplementationIdController2_0Test.java (hunk 2 above)
Comment (BartlomiejRasztabiga) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/ImplementationIdController2_0Test.java]: True :) https://github.com/openmrs/openmrs-module-webservices.rest/pull/496/commits/3fceee7c8a2feba27108597967e39f40f54b0789

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/ImplementationIdController2_0.java (hunk 3 above)
Comment (BartlomiejRasztabiga) [reply omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/ImplementationIdController2_0.java]: I've changed that to not catch this exception and let it be sent to the user: https://github.com/openmrs/openmrs-module-webservices.rest/pull/496/commits/3fa13610c6e0e24de4a1abc9d9914d1101bf9408

File: omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/ImplementationIdController2_0Test.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+		implementationId.setName("name");
+		implementationId.setDescription("description");
+		when(administrationService.getImplementationId()).thenReturn(implementationId);
+
+		// make GET call
+		MockHttpServletRequest req = request(RequestMethod.GET, getURI());
+		SimpleObject result = deserialize(handle(req));
+
+		assertEquals("name", PropertyUtils.getProperty(result, "name"));
+		assertEquals("test", PropertyUtils.getProperty(result, "implementationId"));
```
Comment (dkayiwa) [inline omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/ImplementationIdController2_0Test.java:58]: Why use `test`? Wouldn't it better to use `implementationId` and `passphrase`?

Comment (BartlomiejRasztabiga) [reply omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/ImplementationIdController2_0Test.java]: https://github.com/openmrs/openmrs-module-webservices.rest/pull/496/commits/f06f1bc2932aa4824b79275d03b61c38c9296530

**Pull request merged**